### Note

Please add 'gds_multiindex' file before you run this colab

In [ ]:
import pandas as pd

### Time-invariant panel

Transforming time-invariant variables into a panel format

In [ ]:
ti_data = pd.read_excel('/content/gds_multiindex.xlsx',sheet_name='Ticker-sector')

In [ ]:
# =============================================================================
# time invariant variables: Name, Sector, Tickers etc
# =============================================================================
# Get Ticker and Sector columns
ti=ti_data[['Ticker','Sector']].copy()
# Set Ticker as the index
ti.index=ti['Ticker']
ti.drop('Ticker',axis=1,inplace=True)
ti

Sector
Ticker                         
CIGI.TO             Real Estate
ETN                 Industrails
WOR.AX                   Energy
PFE                 Health Care
CPG.L    Consumer Discretionary
...                         ...
PRY.MI              Industrails
ATOS.PA  Information Technology
ALV      Consumer Discretionary
ATEA.OL  Consumer Discretionary
ETSY.O              Industrails

[100 rows x 1 columns]

In [ ]:
# Transpose the df
ti=ti.T
ti

Ticker      CIGI.TO          ETN  WOR.AX          PFE                   CPG.L  \
Sector  Real Estate  Industrails  Energy  Health Care  Consumer Discretionary   

Ticker                  TIGO.O    DEMANT.CO                     PVH  \
Sector  Communication Services  Health Care  Consumer Discretionary   

Ticker                     IPG             AD.AS  ...                   PSN.L  \
Sector  Communication Services  Consumer Staples  ...  Consumer Discretionary   

Ticker                 CCEP.O                  ADSK.O     ELUXb.ST  \
Sector  Consumer Noncyclicals  Consumer Discretionary  Industrails   

Ticker                    FMC       PRY.MI                 ATOS.PA  \
Sector  Consumer Noncyclicals  Industrails  Information Technology   

Ticker                     ALV                 ATEA.OL       ETSY.O  
Sector  Consumer Discretionary  Consumer Discretionary  Industrails  

[1 rows x 100 columns]

In [ ]:
# Duplicate the df
ti_re=pd.concat([ti]*4).reset_index(drop=True)
ti_re

Ticker      CIGI.TO          ETN  WOR.AX          PFE                   CPG.L  \
0       Real Estate  Industrails  Energy  Health Care  Consumer Discretionary   
1       Real Estate  Industrails  Energy  Health Care  Consumer Discretionary   
2       Real Estate  Industrails  Energy  Health Care  Consumer Discretionary   
3       Real Estate  Industrails  Energy  Health Care  Consumer Discretionary   

Ticker                  TIGO.O    DEMANT.CO                     PVH  \
0       Communication Services  Health Care  Consumer Discretionary   
1       Communication Services  Health Care  Consumer Discretionary   
2       Communication Services  Health Care  Consumer Discretionary   
3       Communication Services  Health Care  Consumer Discretionary   

Ticker                     IPG             AD.AS  ...                   PSN.L  \
0       Communication Services  Consumer Staples  ...  Consumer Discretionary   
1       Communication Services  Consumer Staples  ...  Consumer Discretionary   
2       Communication Services  Consumer Staples  ...  Consumer Discretionary   
3       Communication Services  Consumer Staples  ...  Consumer Discretionary   

Ticker                 CCEP.O                  ADSK.O     ELUXb.ST  \
0       Consumer Noncyclicals  Consumer Discretionary  Industrails   
1       Consumer Noncyclicals  Consumer Discretionary  Industrails   
2       Consumer Noncyclicals  Consumer Discretionary  Industrails   
3       Consumer Noncyclicals  Consumer Discretionary  Industrails   

Ticker                    FMC       PRY.MI                 ATOS.PA  \
0       Consumer Noncyclicals  Industrails  Information Technology   
1       Consumer Noncyclicals  Industrails  Information Technology   
2       Consumer Noncyclicals  Industrails  Information Technology   
3       Consumer Noncyclicals  Industrails  Information Technology   

Ticker                     ALV                 ATEA.OL       ETSY.O  
0       Consumer Discretionary  Consumer Discretionary  Industrails  
1       Consumer Discretionary  Consumer Discretionary  Industrails  
2       Consumer Discretionary  Consumer Discretionary  Industrails  
3       Consumer Discretionary  Consumer Discretionary  Industrails  

[4 rows x 100 columns]

In [ ]:
from datetime import datetime
import numpy as np
# format sample period: start and end dates
# function: string to datetime
start_date = datetime.strptime("2018-01-01", "%Y-%m-%d")
end_date = datetime.strptime("2022-01-01", "%Y-%m-%d")

In [ ]:
# generate a yearly dates series
# format it as df
date_list = pd.date_range(start_date, end_date, freq='Y')
dates=pd.DataFrame(date_list,columns=['Dates'])
dates.reset_index(drop=True, inplace=True)
dates

Dates
0 2018-12-31
1 2019-12-31
2 2020-12-31
3 2021-12-31

In [ ]:
# concat two dfs on the common column
ti_re=pd.concat([dates,ti_re],axis=1)
ti_re=ti_re.set_index('Dates')
ti_re

CIGI.TO          ETN  WOR.AX          PFE  \
Dates                                                       
2018-12-31  Real Estate  Industrails  Energy  Health Care   
2019-12-31  Real Estate  Industrails  Energy  Health Care   
2020-12-31  Real Estate  Industrails  Energy  Health Care   
2021-12-31  Real Estate  Industrails  Energy  Health Care   

                             CPG.L                  TIGO.O    DEMANT.CO  \
Dates                                                                     
2018-12-31  Consumer Discretionary  Communication Services  Health Care   
2019-12-31  Consumer Discretionary  Communication Services  Health Care   
2020-12-31  Consumer Discretionary  Communication Services  Health Care   
2021-12-31  Consumer Discretionary  Communication Services  Health Care   

                               PVH                     IPG             AD.AS  \
Dates                                                                          
2018-12-31  Consumer Discretionary  Communication Services  Consumer Staples   
2019-12-31  Consumer Discretionary  Communication Services  Consumer Staples   
2020-12-31  Consumer Discretionary  Communication Services  Consumer Staples   
2021-12-31  Consumer Discretionary  Communication Services  Consumer Staples   

            ...                   PSN.L                 CCEP.O  \
Dates       ...                                                  
2018-12-31  ...  Consumer Discretionary  Consumer Noncyclicals   
2019-12-31  ...  Consumer Discretionary  Consumer Noncyclicals   
2020-12-31  ...  Consumer Discretionary  Consumer Noncyclicals   
2021-12-31  ...  Consumer Discretionary  Consumer Noncyclicals   

                            ADSK.O     ELUXb.ST                    FMC  \
Dates                                                                    
2018-12-31  Consumer Discretionary  Industrails  Consumer Noncyclicals   
2019-12-31  Consumer Discretionary  Industrails  Consumer Noncyclicals   
2020-12-31  Consumer Discretionary  Industrails  Consumer Noncyclicals   
2021-12-31  Consumer Discretionary  Industrails  Consumer Noncyclicals   

                 PRY.MI                 ATOS.PA                     ALV  \
Dates                                                                     
2018-12-31  Industrails  Information Technology  Consumer Discretionary   
2019-12-31  Industrails  Information Technology  Consumer Discretionary   
2020-12-31  Industrails  Information Technology  Consumer Discretionary   
2021-12-31  Industrails  Information Technology  Consumer Discretionary   

                           ATEA.OL       ETSY.O  
Dates                                            
2018-12-31  Consumer Discretionary  Industrails  
2019-12-31  Consumer Discretionary  Industrails  
2020-12-31  Consumer Discretionary  Industrails  
2021-12-31  Consumer Discretionary  Industrails  

[4 rows x 100 columns]

In [ ]:
# make it into panel format
# name the sector col as Sector
ti_panel=pd.DataFrame(ti_re.unstack())
ti_panel.columns=['Sector']
ti_panel

Sector
        Dates                             
CIGI.TO 2018-12-31             Real Estate
        2019-12-31             Real Estate
        2020-12-31             Real Estate
        2021-12-31             Real Estate
ETN     2018-12-31             Industrails
...                                    ...
ATEA.OL 2021-12-31  Consumer Discretionary
ETSY.O  2018-12-31             Industrails
        2019-12-31             Industrails
        2020-12-31             Industrails
        2021-12-31             Industrails

[400 rows x 1 columns]

In [ ]:
# format the panel & MultiIndex:
# rename the MultiIndex
ti_panel.rename_axis(index=['id','Dates'],inplace=True)
# sort values by MultiIndex
ti_panel=ti_panel.sort_values(['id', 'Dates'], ascending=[True, True])
ti_panel

Sector
id        Dates                             
066570.KS 2018-12-31  Consumer Discretionary
          2019-12-31  Consumer Discretionary
          2020-12-31  Consumer Discretionary
          2021-12-31  Consumer Discretionary
1928.T    2018-12-31  Consumer Discretionary
...                                      ...
WOR.AX    2021-12-31                  Energy
YUMC.K    2018-12-31  Consumer Discretionary
          2019-12-31  Consumer Discretionary
          2020-12-31  Consumer Discretionary
          2021-12-31  Consumer Discretionary

[400 rows x 1 columns]

In [ ]:
# check the # of firms
ti_panel.index.get_level_values(0).nunique()

100

In [ ]:
# check the # of dates
ti_panel.index.get_level_values(1).nunique()

4

In [ ]:
# save the formatted panel into parquet
ti_panel.to_parquet('ti_panel.parquet')

Financial Data Panel

In [ ]:
# =============================================================================
# # FINANCIAL DATA PANEL
# =============================================================================
data=pd.read_excel('/content/gds_multiindex.xlsx',sheet_name='Financials')
data.head()

Ticker             066570.KS     Unnamed: 2 Unnamed: 3  \
0                Dates            Capex Data      EBIT Data   EPS Data   
1  2017-12-31 00:00:00         3219314000000  2468549000000        422   
2  2018-12-31 00:00:00  3844370999999.990234  2703291000000       9579   
3  2019-12-31 00:00:00         2544599000000  2436139000000       6882   
4  2020-12-31 00:00:00         3083042000000  3905108000000        169   

       Unnamed: 4 Unnamed: 5       1928.T    Unnamed: 7 Unnamed: 8  \
0         OE Data   ROA Data   Capex Data     EBIT Data   EPS Data   
1  54029270000000      0.207  94588000000  184164000000     175.48   
2  58927735250000      4.365  62899000000  195540000000     193.06   
3  58638373000000      2.899  53877000000  189223000000   186.5299   
4  59870036000000       0.28  66622000000  205256000000     205.79   

      Unnamed: 9  ...         CWK Unnamed: 492 Unnamed: 493 Unnamed: 494  \
0        OE Data  ...  Capex Data    EBIT Data     EPS Data      OE Data   
1              0  ...   415000000    761000000         1.28   6112000000   
2  1969382000000  ...   470000000    870000000         1.42   6359000000   
3  1971093000000  ...   435000000    926000000         1.53   7462000000   
4   272780000000  ...   419000000    702000000         1.88   7850000000   

  Unnamed: 495    ENGIE.PA Unnamed: 497 Unnamed: 498 Unnamed: 499 Unnamed: 500  
0     ROA Data  Capex Data    EBIT Data     EPS Data      OE Data     ROA Data  
1        13.63   783000000    464000000         0.73   3737000000         0.73  
2        14.12   780000000    424000000         0.85   2255000000         0.89  
3        13.66   907000000    431000000         -0.1            0         -0.1  
4        12.61   824000000    250000000         1.48            0         1.29  

[5 rows x 501 columns]

In [ ]:
# daeling with 'Unnamed' columns
# forward fill, if starts with 'Unnamed', fill it with nan
data.columns = data.columns.to_series().mask(lambda x: x.str.startswith('Unnamed')).ffill()
data.head()

Ticker             066570.KS      066570.KS 066570.KS  \
0                Dates            Capex Data      EBIT Data  EPS Data   
1  2017-12-31 00:00:00         3219314000000  2468549000000       422   
2  2018-12-31 00:00:00  3844370999999.990234  2703291000000      9579   
3  2019-12-31 00:00:00         2544599000000  2436139000000      6882   
4  2020-12-31 00:00:00         3083042000000  3905108000000       169   

        066570.KS 066570.KS       1928.T        1928.T    1928.T  \
0         OE Data  ROA Data   Capex Data     EBIT Data  EPS Data   
1  54029270000000     0.207  94588000000  184164000000    175.48   
2  58927735250000     4.365  62899000000  195540000000    193.06   
3  58638373000000     2.899  53877000000  189223000000  186.5299   
4  59870036000000      0.28  66622000000  205256000000    205.79   

          1928.T  ...         CWK        CWK       CWK         CWK       CWK  \
0        OE Data  ...  Capex Data  EBIT Data  EPS Data     OE Data  ROA Data   
1              0  ...   415000000  761000000      1.28  6112000000     13.63   
2  1969382000000  ...   470000000  870000000      1.42  6359000000     14.12   
3  1971093000000  ...   435000000  926000000      1.53  7462000000     13.66   
4   272780000000  ...   419000000  702000000      1.88  7850000000     12.61   

     ENGIE.PA   ENGIE.PA  ENGIE.PA    ENGIE.PA  ENGIE.PA  
0  Capex Data  EBIT Data  EPS Data     OE Data  ROA Data  
1   783000000  464000000      0.73  3737000000      0.73  
2   780000000  424000000      0.85  2255000000      0.89  
3   907000000  431000000      -0.1           0      -0.1  
4   824000000  250000000      1.48           0      1.29  

[5 rows x 501 columns]

In [ ]:
# name the first item in the column header as 'Dates'
data.columns.values[0] = 'Dates'
data.head()

Dates             066570.KS      066570.KS 066570.KS  \
0                Dates            Capex Data      EBIT Data  EPS Data   
1  2017-12-31 00:00:00         3219314000000  2468549000000       422   
2  2018-12-31 00:00:00  3844370999999.990234  2703291000000      9579   
3  2019-12-31 00:00:00         2544599000000  2436139000000      6882   
4  2020-12-31 00:00:00         3083042000000  3905108000000       169   

        066570.KS 066570.KS       1928.T        1928.T    1928.T  \
0         OE Data  ROA Data   Capex Data     EBIT Data  EPS Data   
1  54029270000000     0.207  94588000000  184164000000    175.48   
2  58927735250000     4.365  62899000000  195540000000    193.06   
3  58638373000000     2.899  53877000000  189223000000  186.5299   
4  59870036000000      0.28  66622000000  205256000000    205.79   

          1928.T  ...         CWK        CWK       CWK         CWK       CWK  \
0        OE Data  ...  Capex Data  EBIT Data  EPS Data     OE Data  ROA Data   
1              0  ...   415000000  761000000      1.28  6112000000     13.63   
2  1969382000000  ...   470000000  870000000      1.42  6359000000     14.12   
3  1971093000000  ...   435000000  926000000      1.53  7462000000     13.66   
4   272780000000  ...   419000000  702000000      1.88  7850000000     12.61   

     ENGIE.PA   ENGIE.PA  ENGIE.PA    ENGIE.PA  ENGIE.PA  
0  Capex Data  EBIT Data  EPS Data     OE Data  ROA Data  
1   783000000  464000000      0.73  3737000000      0.73  
2   780000000  424000000      0.85  2255000000      0.89  
3   907000000  431000000      -0.1           0      -0.1  
4   824000000  250000000      1.48           0      1.29  

[5 rows x 501 columns]

In [ ]:
# get MultiIndex columns:
# make a copy of the original df
df=data.copy()
df.head()

Dates             066570.KS      066570.KS 066570.KS  \
0                Dates            Capex Data      EBIT Data  EPS Data   
1  2017-12-31 00:00:00         3219314000000  2468549000000       422   
2  2018-12-31 00:00:00  3844370999999.990234  2703291000000      9579   
3  2019-12-31 00:00:00         2544599000000  2436139000000      6882   
4  2020-12-31 00:00:00         3083042000000  3905108000000       169   

        066570.KS 066570.KS       1928.T        1928.T    1928.T  \
0         OE Data  ROA Data   Capex Data     EBIT Data  EPS Data   
1  54029270000000     0.207  94588000000  184164000000    175.48   
2  58927735250000     4.365  62899000000  195540000000    193.06   
3  58638373000000     2.899  53877000000  189223000000  186.5299   
4  59870036000000      0.28  66622000000  205256000000    205.79   

          1928.T  ...         CWK        CWK       CWK         CWK       CWK  \
0        OE Data  ...  Capex Data  EBIT Data  EPS Data     OE Data  ROA Data   
1              0  ...   415000000  761000000      1.28  6112000000     13.63   
2  1969382000000  ...   470000000  870000000      1.42  6359000000     14.12   
3  1971093000000  ...   435000000  926000000      1.53  7462000000     13.66   
4   272780000000  ...   419000000  702000000      1.88  7850000000     12.61   

     ENGIE.PA   ENGIE.PA  ENGIE.PA    ENGIE.PA  ENGIE.PA  
0  Capex Data  EBIT Data  EPS Data     OE Data  ROA Data  
1   783000000  464000000      0.73  3737000000      0.73  
2   780000000  424000000      0.85  2255000000      0.89  
3   907000000  431000000      -0.1           0      -0.1  
4   824000000  250000000      1.48           0      1.29  

[5 rows x 501 columns]

In [ ]:
# drop 'Dates' column
df.drop(['Dates'],axis=1,inplace=True)
df.head()

066570.KS      066570.KS 066570.KS       066570.KS 066570.KS  \
0            Capex Data      EBIT Data  EPS Data         OE Data  ROA Data   
1         3219314000000  2468549000000       422  54029270000000     0.207   
2  3844370999999.990234  2703291000000      9579  58927735250000     4.365   
3         2544599000000  2436139000000      6882  58638373000000     2.899   
4         3083042000000  3905108000000       169  59870036000000      0.28   

        1928.T        1928.T    1928.T         1928.T    1928.T  ...  \
0   Capex Data     EBIT Data  EPS Data        OE Data  ROA Data  ...   
1  94588000000  184164000000    175.48              0       9.1  ...   
2  62899000000  195540000000    193.06  1969382000000       8.8  ...   
3  53877000000  189223000000  186.5299  1971093000000     5.329  ...   
4  66622000000  205256000000    205.79   272780000000       8.5  ...   

          CWK        CWK       CWK         CWK       CWK    ENGIE.PA  \
0  Capex Data  EBIT Data  EPS Data     OE Data  ROA Data  Capex Data   
1   415000000  761000000      1.28  6112000000     13.63   783000000   
2   470000000  870000000      1.42  6359000000     14.12   780000000   
3   435000000  926000000      1.53  7462000000     13.66   907000000   
4   419000000  702000000      1.88  7850000000     12.61   824000000   

    ENGIE.PA  ENGIE.PA    ENGIE.PA  ENGIE.PA  
0  EBIT Data  EPS Data     OE Data  ROA Data  
1  464000000      0.73  3737000000      0.73  
2  424000000      0.85  2255000000      0.89  
3  431000000      -0.1           0      -0.1  
4  250000000      1.48           0      1.29  

[5 rows x 500 columns]

In [ ]:
# get MultiIndex from arrays of df columns [Tickers] and the name of data fields in the first row [EBIT, BS_TOT_ASSET ...]
mul_cols=pd.MultiIndex.from_arrays([df.columns, df.iloc[0].values])
mul_cols

MultiIndex([('066570.KS', 'Capex Data'),
            ('066570.KS',  'EBIT Data'),
            ('066570.KS',   'EPS Data'),
            ('066570.KS',    'OE Data'),
            ('066570.KS',   'ROA Data'),
            (   '1928.T', 'Capex Data'),
            (   '1928.T',  'EBIT Data'),
            (   '1928.T',   'EPS Data'),
            (   '1928.T',    'OE Data'),
            (   '1928.T',   'ROA Data'),
            ...
            (      'CWK', 'Capex Data'),
            (      'CWK',  'EBIT Data'),
            (      'CWK',   'EPS Data'),
            (      'CWK',    'OE Data'),
            (      'CWK',   'ROA Data'),
            ( 'ENGIE.PA', 'Capex Data'),
            ( 'ENGIE.PA',  'EBIT Data'),
            ( 'ENGIE.PA',   'EPS Data'),
            ( 'ENGIE.PA',    'OE Data'),
            ( 'ENGIE.PA',   'ROA Data')],
           length=500)

In [ ]:
# get values only df
data.head()

Dates             066570.KS      066570.KS 066570.KS  \
0                Dates            Capex Data      EBIT Data  EPS Data   
1  2017-12-31 00:00:00         3219314000000  2468549000000       422   
2  2018-12-31 00:00:00  3844370999999.990234  2703291000000      9579   
3  2019-12-31 00:00:00         2544599000000  2436139000000      6882   
4  2020-12-31 00:00:00         3083042000000  3905108000000       169   

        066570.KS 066570.KS       1928.T        1928.T    1928.T  \
0         OE Data  ROA Data   Capex Data     EBIT Data  EPS Data   
1  54029270000000     0.207  94588000000  184164000000    175.48   
2  58927735250000     4.365  62899000000  195540000000    193.06   
3  58638373000000     2.899  53877000000  189223000000  186.5299   
4  59870036000000      0.28  66622000000  205256000000    205.79   

          1928.T  ...         CWK        CWK       CWK         CWK       CWK  \
0        OE Data  ...  Capex Data  EBIT Data  EPS Data     OE Data  ROA Data   
1              0  ...   415000000  761000000      1.28  6112000000     13.63   
2  1969382000000  ...   470000000  870000000      1.42  6359000000     14.12   
3  1971093000000  ...   435000000  926000000      1.53  7462000000     13.66   
4   272780000000  ...   419000000  702000000      1.88  7850000000     12.61   

     ENGIE.PA   ENGIE.PA  ENGIE.PA    ENGIE.PA  ENGIE.PA  
0  Capex Data  EBIT Data  EPS Data     OE Data  ROA Data  
1   783000000  464000000      0.73  3737000000      0.73  
2   780000000  424000000      0.85  2255000000      0.89  
3   907000000  431000000      -0.1           0      -0.1  
4   824000000  250000000      1.48           0      1.29  

[5 rows x 501 columns]

In [ ]:
# drop the first row and reset index
data=data.drop([0]).reset_index(drop=True)
data

Dates             066570.KS      066570.KS 066570.KS  \
0  2017-12-31 00:00:00         3219314000000  2468549000000       422   
1  2018-12-31 00:00:00  3844370999999.990234  2703291000000      9579   
2  2019-12-31 00:00:00         2544599000000  2436139000000      6882   
3  2020-12-31 00:00:00         3083042000000  3905108000000       169   
4  2021-12-31 00:00:00         3259091000000  4057997000000     10926   

        066570.KS 066570.KS       1928.T        1928.T    1928.T  \
0  54029270000000     0.207  94588000000  184164000000    175.48   
1  58927735250000     4.365  62899000000  195540000000    193.06   
2  58638373000000     2.899  53877000000  189223000000  186.5299   
3  59870036000000      0.28  66622000000  205256000000    205.79   
4  60067059000000      4.23  87490000000  186519000000    181.18   

          1928.T  ...        CWK        CWK   CWK         CWK    CWK  \
0              0  ...  415000000  761000000  1.28  6112000000  13.63   
1  1969382000000  ...  470000000  870000000  1.42  6359000000  14.12   
2  1971093000000  ...  435000000  926000000  1.53  7462000000  13.66   
3   272780000000  ...  419000000  702000000  1.88  7850000000  12.61   
4  2260385000000  ...  425000000  834000000  1.73  7230000000  13.62   

    ENGIE.PA   ENGIE.PA ENGIE.PA    ENGIE.PA ENGIE.PA  
0  783000000  464000000     0.73  3737000000     0.73  
1  780000000  424000000     0.85  2255000000     0.89  
2  907000000  431000000     -0.1           0     -0.1  
3  824000000  250000000     1.48           0     1.29  
4  824000000  350000000     1.56           0     1.15  

[5 rows x 501 columns]

In [ ]:
# set 'Dates' to datetime format
data['Dates']=pd.to_datetime(data['Dates'])
data

Dates             066570.KS      066570.KS 066570.KS       066570.KS  \
0 2017-12-31         3219314000000  2468549000000       422  54029270000000   
1 2018-12-31  3844370999999.990234  2703291000000      9579  58927735250000   
2 2019-12-31         2544599000000  2436139000000      6882  58638373000000   
3 2020-12-31         3083042000000  3905108000000       169  59870036000000   
4 2021-12-31         3259091000000  4057997000000     10926  60067059000000   

  066570.KS       1928.T        1928.T    1928.T         1928.T  ...  \
0     0.207  94588000000  184164000000    175.48              0  ...   
1     4.365  62899000000  195540000000    193.06  1969382000000  ...   
2     2.899  53877000000  189223000000  186.5299  1971093000000  ...   
3      0.28  66622000000  205256000000    205.79   272780000000  ...   
4      4.23  87490000000  186519000000    181.18  2260385000000  ...   

         CWK        CWK   CWK         CWK    CWK   ENGIE.PA   ENGIE.PA  \
0  415000000  761000000  1.28  6112000000  13.63  783000000  464000000   
1  470000000  870000000  1.42  6359000000  14.12  780000000  424000000   
2  435000000  926000000  1.53  7462000000  13.66  907000000  431000000   
3  419000000  702000000  1.88  7850000000  12.61  824000000  250000000   
4  425000000  834000000  1.73  7230000000  13.62  824000000  350000000   

  ENGIE.PA    ENGIE.PA ENGIE.PA  
0     0.73  3737000000     0.73  
1     0.85  2255000000     0.89  
2     -0.1           0     -0.1  
3     1.48           0     1.29  
4     1.56           0     1.15  

[5 rows x 501 columns]

In [ ]:
# set 'Dates' as df index
data.set_index(data['Dates'], inplace=True)
data.drop(['Dates'],axis=1,inplace=True)
data

066570.KS      066570.KS 066570.KS       066570.KS  \
Dates                                                                       
2017-12-31         3219314000000  2468549000000       422  54029270000000   
2018-12-31  3844370999999.990234  2703291000000      9579  58927735250000   
2019-12-31         2544599000000  2436139000000      6882  58638373000000   
2020-12-31         3083042000000  3905108000000       169  59870036000000   
2021-12-31         3259091000000  4057997000000     10926  60067059000000   

           066570.KS       1928.T        1928.T    1928.T         1928.T  \
Dates                                                                      
2017-12-31     0.207  94588000000  184164000000    175.48              0   
2018-12-31     4.365  62899000000  195540000000    193.06  1969382000000   
2019-12-31     2.899  53877000000  189223000000  186.5299  1971093000000   
2020-12-31      0.28  66622000000  205256000000    205.79   272780000000   
2021-12-31      4.23  87490000000  186519000000    181.18  2260385000000   

           1928.T  ...        CWK        CWK   CWK         CWK    CWK  \
Dates              ...                                                  
2017-12-31    9.1  ...  415000000  761000000  1.28  6112000000  13.63   
2018-12-31    8.8  ...  470000000  870000000  1.42  6359000000  14.12   
2019-12-31  5.329  ...  435000000  926000000  1.53  7462000000  13.66   
2020-12-31    8.5  ...  419000000  702000000  1.88  7850000000  12.61   
2021-12-31      7  ...  425000000  834000000  1.73  7230000000  13.62   

             ENGIE.PA   ENGIE.PA ENGIE.PA    ENGIE.PA ENGIE.PA  
Dates                                                           
2017-12-31  783000000  464000000     0.73  3737000000     0.73  
2018-12-31  780000000  424000000     0.85  2255000000     0.89  
2019-12-31  907000000  431000000     -0.1           0     -0.1  
2020-12-31  824000000  250000000     1.48           0     1.29  
2021-12-31  824000000  350000000     1.56           0     1.15  

[5 rows x 500 columns]

In [ ]:
# add previously created mul_cols as data df's columns
data.columns=mul_cols
data.head()

066570.KS                                          \
                      Capex Data      EBIT Data EPS Data         OE Data   
Dates                                                                      
2017-12-31         3219314000000  2468549000000      422  54029270000000   
2018-12-31  3844370999999.990234  2703291000000     9579  58927735250000   
2019-12-31         2544599000000  2436139000000     6882  58638373000000   
2020-12-31         3083042000000  3905108000000      169  59870036000000   
2021-12-31         3259091000000  4057997000000    10926  60067059000000   

                          1928.T                                         \
           ROA Data   Capex Data     EBIT Data  EPS Data        OE Data   
Dates                                                                     
2017-12-31    0.207  94588000000  184164000000    175.48              0   
2018-12-31    4.365  62899000000  195540000000    193.06  1969382000000   
2019-12-31    2.899  53877000000  189223000000  186.5299  1971093000000   
2020-12-31     0.28  66622000000  205256000000    205.79   272780000000   
2021-12-31     4.23  87490000000  186519000000    181.18  2260385000000   

                     ...        CWK                                           \
           ROA Data  ... Capex Data  EBIT Data EPS Data     OE Data ROA Data   
Dates                ...                                                       
2017-12-31      9.1  ...  415000000  761000000     1.28  6112000000    13.63   
2018-12-31      8.8  ...  470000000  870000000     1.42  6359000000    14.12   
2019-12-31    5.329  ...  435000000  926000000     1.53  7462000000    13.66   
2020-12-31      8.5  ...  419000000  702000000     1.88  7850000000    12.61   
2021-12-31        7  ...  425000000  834000000     1.73  7230000000    13.62   

             ENGIE.PA                                           
           Capex Data  EBIT Data EPS Data     OE Data ROA Data  
Dates                                                           
2017-12-31  783000000  464000000     0.73  3737000000     0.73  
2018-12-31  780000000  424000000     0.85  2255000000     0.89  
2019-12-31  907000000  431000000     -0.1           0     -0.1  
2020-12-31  824000000  250000000     1.48           0     1.29  
2021-12-31  824000000  350000000     1.56           0     1.15  

[5 rows x 500 columns]

In [ ]:
# before convert the data df (with multi-level column headers) into a panel
# shift the dates index! becuase we want to use lagged financial data to explain the DV
# check the current datetime index
data.index

DatetimeIndex(['2017-12-31', '2018-12-31', '2019-12-31', '2020-12-31',
               '2021-12-31'],
              dtype='datetime64[ns]', name='Dates', freq=None)

In [ ]:
# shifting the original datetime index one-year forward
data.index = data.index + pd.tseries.offsets.YearEnd()
data.index

DatetimeIndex(['2018-12-31', '2019-12-31', '2020-12-31', '2021-12-31',
               '2022-12-31'],
              dtype='datetime64[ns]', name='Dates', freq=None)

In [ ]:
# slice the df based on datetime index
data=data[:'2021-12-31']
data

066570.KS                                          \
                      Capex Data      EBIT Data EPS Data         OE Data   
Dates                                                                      
2018-12-31         3219314000000  2468549000000      422  54029270000000   
2019-12-31  3844370999999.990234  2703291000000     9579  58927735250000   
2020-12-31         2544599000000  2436139000000     6882  58638373000000   
2021-12-31         3083042000000  3905108000000      169  59870036000000   

                          1928.T                                         \
           ROA Data   Capex Data     EBIT Data  EPS Data        OE Data   
Dates                                                                     
2018-12-31    0.207  94588000000  184164000000    175.48              0   
2019-12-31    4.365  62899000000  195540000000    193.06  1969382000000   
2020-12-31    2.899  53877000000  189223000000  186.5299  1971093000000   
2021-12-31     0.28  66622000000  205256000000    205.79   272780000000   

                     ...        CWK                                           \
           ROA Data  ... Capex Data  EBIT Data EPS Data     OE Data ROA Data   
Dates                ...                                                       
2018-12-31      9.1  ...  415000000  761000000     1.28  6112000000    13.63   
2019-12-31      8.8  ...  470000000  870000000     1.42  6359000000    14.12   
2020-12-31    5.329  ...  435000000  926000000     1.53  7462000000    13.66   
2021-12-31      8.5  ...  419000000  702000000     1.88  7850000000    12.61   

             ENGIE.PA                                           
           Capex Data  EBIT Data EPS Data     OE Data ROA Data  
Dates                                                           
2018-12-31  783000000  464000000     0.73  3737000000     0.73  
2019-12-31  780000000  424000000     0.85  2255000000     0.89  
2020-12-31  907000000  431000000     -0.1           0     -0.1  
2021-12-31  824000000  250000000     1.48           0     1.29  

[4 rows x 500 columns]

In [ ]:
# convert the data df (with Multi-level columns) into panel
fin_panel=data.stack(0,dropna=False)
fin_panel.head()

Capex Data      EBIT Data EPS Data         OE Data  \
Dates                                                                         
2018-12-31 066570.KS  3219314000000  2468549000000      422  54029270000000   
           1928.T       94588000000   184164000000   175.48               0   
           2502.T       87882000000   188248000000   194.75   1570012000000   
           2802.T       86193000000    82662000000    91.99   1006075000000   
           3289.T       53336000000    73227000000    51.77               0   

                     ROA Data  
Dates                          
2018-12-31 066570.KS    0.207  
           1928.T         9.1  
           2502.T       4.276  
           2802.T         6.9  
           3289.T         3.1

In [ ]:
# rename the MultiIndex
fin_panel.rename_axis(index=['Dates', 'id'],inplace=True)
# swap the levels of index, since we want to have id-dates like index
fin_panel=fin_panel.swaplevel()
# sort values by MultiIndex
fin_panel=fin_panel.sort_values(['id', 'Dates'], ascending=[True, True])
# set the df value format as float
fin_panel=fin_panel.astype(float)
fin_panel.head()

Capex Data     EBIT Data  EPS Data       OE Data  \
id        Dates                                                            
066570.KS 2018-12-31  3.219314e+12  2.468549e+12    422.00  5.402927e+13   
          2019-12-31  3.844371e+12  2.703291e+12   9579.00  5.892774e+13   
          2020-12-31  2.544599e+12  2.436139e+12   6882.00  5.863837e+13   
          2021-12-31  3.083042e+12  3.905108e+12    169.00  5.987004e+13   
1928.T    2018-12-31  9.458800e+10  1.841640e+11    175.48  0.000000e+00   

                      ROA Data  
id        Dates                 
066570.KS 2018-12-31     0.207  
          2019-12-31     4.365  
          2020-12-31     2.899  
          2021-12-31     0.280  
1928.T    2018-12-31     9.100

In [ ]:
# save formatted financial data panel into parquet
fin_panel.to_parquet('fin_panel.parquet')

### Emmissions data Panel

In [ ]:
# =============================================================================
# # EMISSION DATA PANEL
# =============================================================================
# emission data to panel format
emission=pd.read_excel('/content/gds_multiindex.xlsx',sheet_name='Emmissions')
emission.head()

Ticker 066570.KS        Unnamed: 2      Unnamed: 3   1928.T  \
0                Dates   Scope 1  Scope 2 Location  Scope 2 Market  Scope 1   
1  2018-12-31 00:00:00    785312            846510            5226    72967   
2  2019-12-31 00:00:00    611675            891246            9114    70574   
3  2020-12-31 00:00:00    440166            854054            5080    85901   
4  2021-12-31 00:00:00    332724            847031          819240    52917   

         Unnamed: 5      Unnamed: 6   2502.T        Unnamed: 8  \
0  Scope 2 Location  Scope 2 Market  Scope 1  Scope 2 Location   
1             59606           58259   462598            423401   
2             55313           50808   478695            442398   
3             54318           48928   463242            372437   
4             48539           35270   493847            306940   

       Unnamed: 9  ...    Unnamed: 291   YUMC.K      Unnamed: 293  \
0  Scope 2 Market  ...  Scope 2 Market  Scope 1  Scope 2 Location   
1          418724  ...               0        0                 0   
2          430127  ...               0   114720           1956072   
3          366869  ...           42268   135615           1873757   
4          306303  ...           25973   158887           2158856   

     Unnamed: 294      CWK      Unnamed: 296    Unnamed: 297  ENGIE.PA  \
0  Scope 2 Market  Scope 1  Scope 2 Location  Scope 2 Market   Scope 1   
1               0        0                 0               0  57205670   
2               0    15686             21177           21697  46188978   
3               0    15809             20602           21535  38589016   
4               0    15665             17482           18518  37528756   

       Unnamed: 299    Unnamed: 300  
0  Scope 2 Location  Scope 2 Market  
1           2912586         2912586  
2           2534464         2534464  
3           2330625         2330625  
4           1922614         1922614  

[5 rows x 301 columns]

In [ ]:
#  forward fill, if starts with 'Unnamed', fill it with nan
emission.columns = emission.columns.to_series().mask(lambda x: x.str.startswith('Unnamed')).ffill()
emission.head()

Ticker 066570.KS         066570.KS       066570.KS   1928.T  \
0                Dates   Scope 1  Scope 2 Location  Scope 2 Market  Scope 1   
1  2018-12-31 00:00:00    785312            846510            5226    72967   
2  2019-12-31 00:00:00    611675            891246            9114    70574   
3  2020-12-31 00:00:00    440166            854054            5080    85901   
4  2021-12-31 00:00:00    332724            847031          819240    52917   

             1928.T          1928.T   2502.T            2502.T  \
0  Scope 2 Location  Scope 2 Market  Scope 1  Scope 2 Location   
1             59606           58259   462598            423401   
2             55313           50808   478695            442398   
3             54318           48928   463242            372437   
4             48539           35270   493847            306940   

           2502.T  ...          WOR.AX   YUMC.K            YUMC.K  \
0  Scope 2 Market  ...  Scope 2 Market  Scope 1  Scope 2 Location   
1          418724  ...               0        0                 0   
2          430127  ...               0   114720           1956072   
3          366869  ...           42268   135615           1873757   
4          306303  ...           25973   158887           2158856   

           YUMC.K      CWK               CWK             CWK  ENGIE.PA  \
0  Scope 2 Market  Scope 1  Scope 2 Location  Scope 2 Market   Scope 1   
1               0        0                 0               0  57205670   
2               0    15686             21177           21697  46188978   
3               0    15809             20602           21535  38589016   
4               0    15665             17482           18518  37528756   

           ENGIE.PA        ENGIE.PA  
0  Scope 2 Location  Scope 2 Market  
1           2912586         2912586  
2           2534464         2534464  
3           2330625         2330625  
4           1922614         1922614  

[5 rows x 301 columns]

In [ ]:
# get multi-columns
# get a copy of original emissions df
df=emission.copy()
df.head()

Ticker 066570.KS         066570.KS       066570.KS   1928.T  \
0                Dates   Scope 1  Scope 2 Location  Scope 2 Market  Scope 1   
1  2018-12-31 00:00:00    785312            846510            5226    72967   
2  2019-12-31 00:00:00    611675            891246            9114    70574   
3  2020-12-31 00:00:00    440166            854054            5080    85901   
4  2021-12-31 00:00:00    332724            847031          819240    52917   

             1928.T          1928.T   2502.T            2502.T  \
0  Scope 2 Location  Scope 2 Market  Scope 1  Scope 2 Location   
1             59606           58259   462598            423401   
2             55313           50808   478695            442398   
3             54318           48928   463242            372437   
4             48539           35270   493847            306940   

           2502.T  ...          WOR.AX   YUMC.K            YUMC.K  \
0  Scope 2 Market  ...  Scope 2 Market  Scope 1  Scope 2 Location   
1          418724  ...               0        0                 0   
2          430127  ...               0   114720           1956072   
3          366869  ...           42268   135615           1873757   
4          306303  ...           25973   158887           2158856   

           YUMC.K      CWK               CWK             CWK  ENGIE.PA  \
0  Scope 2 Market  Scope 1  Scope 2 Location  Scope 2 Market   Scope 1   
1               0        0                 0               0  57205670   
2               0    15686             21177           21697  46188978   
3               0    15809             20602           21535  38589016   
4               0    15665             17482           18518  37528756   

           ENGIE.PA        ENGIE.PA  
0  Scope 2 Location  Scope 2 Market  
1           2912586         2912586  
2           2534464         2534464  
3           2330625         2330625  
4           1922614         1922614  

[5 rows x 301 columns]

In [ ]:
# set the NaN value in the header as Dates
df.columns.values[0] = 'Dates'
df.head()

Dates 066570.KS         066570.KS       066570.KS   1928.T  \
0                Dates   Scope 1  Scope 2 Location  Scope 2 Market  Scope 1   
1  2018-12-31 00:00:00    785312            846510            5226    72967   
2  2019-12-31 00:00:00    611675            891246            9114    70574   
3  2020-12-31 00:00:00    440166            854054            5080    85901   
4  2021-12-31 00:00:00    332724            847031          819240    52917   

             1928.T          1928.T   2502.T            2502.T  \
0  Scope 2 Location  Scope 2 Market  Scope 1  Scope 2 Location   
1             59606           58259   462598            423401   
2             55313           50808   478695            442398   
3             54318           48928   463242            372437   
4             48539           35270   493847            306940   

           2502.T  ...          WOR.AX   YUMC.K            YUMC.K  \
0  Scope 2 Market  ...  Scope 2 Market  Scope 1  Scope 2 Location   
1          418724  ...               0        0                 0   
2          430127  ...               0   114720           1956072   
3          366869  ...           42268   135615           1873757   
4          306303  ...           25973   158887           2158856   

           YUMC.K      CWK               CWK             CWK  ENGIE.PA  \
0  Scope 2 Market  Scope 1  Scope 2 Location  Scope 2 Market   Scope 1   
1               0        0                 0               0  57205670   
2               0    15686             21177           21697  46188978   
3               0    15809             20602           21535  38589016   
4               0    15665             17482           18518  37528756   

           ENGIE.PA        ENGIE.PA  
0  Scope 2 Location  Scope 2 Market  
1           2912586         2912586  
2           2534464         2534464  
3           2330625         2330625  
4           1922614         1922614  

[5 rows x 301 columns]

In [ ]:
# drop the 'Dates' column
df.drop(['Dates'],axis=1,inplace=True)
# set up two levels of multi columns: df's column - headers [Dates, Ticker 1, Ticker 2...] & the first row [Dates, Scoep 1...]
mul_cols=pd.MultiIndex.from_arrays([df.columns, df.iloc[0].values])
mul_cols

MultiIndex([('066570.KS',          'Scope 1'),
            ('066570.KS', 'Scope 2 Location'),
            ('066570.KS',   'Scope 2 Market'),
            (   '1928.T',          'Scope 1'),
            (   '1928.T', 'Scope 2 Location'),
            (   '1928.T',   'Scope 2 Market'),
            (   '2502.T',          'Scope 1'),
            (   '2502.T', 'Scope 2 Location'),
            (   '2502.T',   'Scope 2 Market'),
            (   '2802.T',          'Scope 1'),
            ...
            (   'WOR.AX',   'Scope 2 Market'),
            (   'YUMC.K',          'Scope 1'),
            (   'YUMC.K', 'Scope 2 Location'),
            (   'YUMC.K',   'Scope 2 Market'),
            (      'CWK',          'Scope 1'),
            (      'CWK', 'Scope 2 Location'),
            (      'CWK',   'Scope 2 Market'),
            ( 'ENGIE.PA',          'Scope 1'),
            ( 'ENGIE.PA', 'Scope 2 Location'),
            ( 'ENGIE.PA',   'Scope 2 Market')],
           length=300)

In [ ]:
# get emission values only df
emission.head()

Dates 066570.KS         066570.KS       066570.KS   1928.T  \
0                Dates   Scope 1  Scope 2 Location  Scope 2 Market  Scope 1   
1  2018-12-31 00:00:00    785312            846510            5226    72967   
2  2019-12-31 00:00:00    611675            891246            9114    70574   
3  2020-12-31 00:00:00    440166            854054            5080    85901   
4  2021-12-31 00:00:00    332724            847031          819240    52917   

             1928.T          1928.T   2502.T            2502.T  \
0  Scope 2 Location  Scope 2 Market  Scope 1  Scope 2 Location   
1             59606           58259   462598            423401   
2             55313           50808   478695            442398   
3             54318           48928   463242            372437   
4             48539           35270   493847            306940   

           2502.T  ...          WOR.AX   YUMC.K            YUMC.K  \
0  Scope 2 Market  ...  Scope 2 Market  Scope 1  Scope 2 Location   
1          418724  ...               0        0                 0   
2          430127  ...               0   114720           1956072   
3          366869  ...           42268   135615           1873757   
4          306303  ...           25973   158887           2158856   

           YUMC.K      CWK               CWK             CWK  ENGIE.PA  \
0  Scope 2 Market  Scope 1  Scope 2 Location  Scope 2 Market   Scope 1   
1               0        0                 0               0  57205670   
2               0    15686             21177           21697  46188978   
3               0    15809             20602           21535  38589016   
4               0    15665             17482           18518  37528756   

           ENGIE.PA        ENGIE.PA  
0  Scope 2 Location  Scope 2 Market  
1           2912586         2912586  
2           2534464         2534464  
3           2330625         2330625  
4           1922614         1922614  

[5 rows x 301 columns]

In [ ]:
# drop the first row and reset the index
emission=emission.drop([0]).reset_index(drop=True)
emission.head()

Dates 066570.KS 066570.KS 066570.KS 1928.T 1928.T 1928.T  \
0  2018-12-31 00:00:00    785312    846510      5226  72967  59606  58259   
1  2019-12-31 00:00:00    611675    891246      9114  70574  55313  50808   
2  2020-12-31 00:00:00    440166    854054      5080  85901  54318  48928   
3  2021-12-31 00:00:00    332724    847031    819240  52917  48539  35270   

   2502.T  2502.T  2502.T  ... WOR.AX  YUMC.K   YUMC.K YUMC.K    CWK    CWK  \
0  462598  423401  418724  ...      0       0        0      0      0      0   
1  478695  442398  430127  ...      0  114720  1956072      0  15686  21177   
2  463242  372437  366869  ...  42268  135615  1873757      0  15809  20602   
3  493847  306940  306303  ...  25973  158887  2158856      0  15665  17482   

     CWK  ENGIE.PA ENGIE.PA ENGIE.PA  
0      0  57205670  2912586  2912586  
1  21697  46188978  2534464  2534464  
2  21535  38589016  2330625  2330625  
3  18518  37528756  1922614  1922614  

[4 rows x 301 columns]

In [ ]:
# format the 'Dates' as datetime
emission['Dates']=pd.to_datetime(emission['Dates'])

In [ ]:
# and set 'Dates' column as index
emission.set_index(emission['Dates'],inplace=True)
emission.head()

Dates 066570.KS 066570.KS 066570.KS 1928.T 1928.T 1928.T  \
Dates                                                                      
2018-12-31 2018-12-31    785312    846510      5226  72967  59606  58259   
2019-12-31 2019-12-31    611675    891246      9114  70574  55313  50808   
2020-12-31 2020-12-31    440166    854054      5080  85901  54318  48928   
2021-12-31 2021-12-31    332724    847031    819240  52917  48539  35270   

            2502.T  2502.T  2502.T  ... WOR.AX  YUMC.K   YUMC.K YUMC.K    CWK  \
Dates                               ...                                         
2018-12-31  462598  423401  418724  ...      0       0        0      0      0   
2019-12-31  478695  442398  430127  ...      0  114720  1956072      0  15686   
2020-12-31  463242  372437  366869  ...  42268  135615  1873757      0  15809   
2021-12-31  493847  306940  306303  ...  25973  158887  2158856      0  15665   

              CWK    CWK  ENGIE.PA ENGIE.PA ENGIE.PA  
Dates                                                 
2018-12-31      0      0  57205670  2912586  2912586  
2019-12-31  21177  21697  46188978  2534464  2534464  
2020-12-31  20602  21535  38589016  2330625  2330625  
2021-12-31  17482  18518  37528756  1922614  1922614  

[4 rows x 301 columns]

In [ ]:
# drop the 'Dates' column
emission.drop(['Dates'],axis=1,inplace=True)

In [ ]:
emission.head()

066570.KS 066570.KS 066570.KS 1928.T 1928.T 1928.T  2502.T  2502.T  \
Dates                                                                           
2018-12-31    785312    846510      5226  72967  59606  58259  462598  423401   
2019-12-31    611675    891246      9114  70574  55313  50808  478695  442398   
2020-12-31    440166    854054      5080  85901  54318  48928  463242  372437   
2021-12-31    332724    847031    819240  52917  48539  35270  493847  306940   

            2502.T   2802.T  ... WOR.AX  YUMC.K   YUMC.K YUMC.K    CWK    CWK  \
Dates                        ...                                                
2018-12-31  418724  1196969  ...      0       0        0      0      0      0   
2019-12-31  430127  1013315  ...      0  114720  1956072      0  15686  21177   
2020-12-31  366869  1008811  ...  42268  135615  1873757      0  15809  20602   
2021-12-31  306303  1005363  ...  25973  158887  2158856      0  15665  17482   

              CWK  ENGIE.PA ENGIE.PA ENGIE.PA  
Dates                                          
2018-12-31      0  57205670  2912586  2912586  
2019-12-31  21697  46188978  2534464  2534464  
2020-12-31  21535  38589016  2330625  2330625  
2021-12-31  18518  37528756  1922614  1922614  

[4 rows x 300 columns]

In [ ]:
# set the previously created multi-level columns (i.e., headers) as columns
emission.columns=mul_cols
emission.head()

066570.KS                                  1928.T                   \
             Scope 1 Scope 2 Location Scope 2 Market Scope 1 Scope 2 Location   
Dates                                                                           
2018-12-31    785312           846510           5226   72967            59606   
2019-12-31    611675           891246           9114   70574            55313   
2020-12-31    440166           854054           5080   85901            54318   
2021-12-31    332724           847031         819240   52917            48539   

                           2502.T                                   2802.T  \
           Scope 2 Market Scope 1 Scope 2 Location Scope 2 Market  Scope 1   
Dates                                                                        
2018-12-31          58259  462598           423401         418724  1196969   
2019-12-31          50808  478695           442398         430127  1013315   
2020-12-31          48928  463242           372437         366869  1008811   
2021-12-31          35270  493847           306940         306303  1005363   

            ...         WOR.AX  YUMC.K                                  \
            ... Scope 2 Market Scope 1 Scope 2 Location Scope 2 Market   
Dates       ...                                                          
2018-12-31  ...              0       0                0              0   
2019-12-31  ...              0  114720          1956072              0   
2020-12-31  ...          42268  135615          1873757              0   
2021-12-31  ...          25973  158887          2158856              0   

               CWK                                  ENGIE.PA                   \
           Scope 1 Scope 2 Location Scope 2 Market   Scope 1 Scope 2 Location   
Dates                                                                           
2018-12-31       0                0              0  57205670          2912586   
2019-12-31   15686            21177          21697  46188978          2534464   
2020-12-31   15809            20602          21535  38589016          2330625   
2021-12-31   15665            17482          18518  37528756          1922614   

                           
           Scope 2 Market  
Dates                      
2018-12-31        2912586  
2019-12-31        2534464  
2020-12-31        2330625  
2021-12-31        1922614  

[4 rows x 300 columns]

In [ ]:
# convert the emission df that with two levels of columns (i.e., headers) into panel format
emission_panel=emission.stack(0,dropna=False)
emission_panel.head()

Scope 1 Scope 2 Location Scope 2 Market
Dates                                                        
2018-12-31 066570.KS   785312           846510           5226
           1928.T       72967            59606          58259
           2502.T      462598           423401         418724
           2802.T     1196969          1026764        1015723
           3289.T           0                0              0

In [ ]:
# rename the MultiIndex
emission_panel.rename_axis(index=['Dates', 'id'],inplace=True)
# swap the index levels
emission_panel=emission_panel.swaplevel()
# sort values by MultiIndex
emission_panel=emission_panel.sort_values(['id', 'Dates'], ascending=[True, True])
#
emission_panel.head()

Scope 1 Scope 2 Location Scope 2 Market
id        Dates                                             
066570.KS 2018-12-31  785312           846510           5226
          2019-12-31  611675           891246           9114
          2020-12-31  440166           854054           5080
          2021-12-31  332724           847031         819240
1928.T    2018-12-31   72967            59606          58259

In [ ]:
# set the values as float
emission_panel=emission_panel.astype(float)

In [ ]:
# save the formatted emission panel to parquet
emission_panel.to_parquet('emission_panel.parquet')

In [ ]:
# =============================================================================
#                    FULL PANEL
# =============================================================================
# load all three panels
ti_panel=pd.read_parquet('ti_panel.parquet')
fin_panel=pd.read_parquet('fin_panel.parquet')
emission_panel=pd.read_parquet('emission_panel.parquet')
# check their info
ti_panel.info()
fin_panel.info()
emission_panel.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 400 entries, ('066570.KS', Timestamp('2018-12-31 00:00:00')) to ('YUMC.K', Timestamp('2021-12-31 00:00:00'))
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Sector  400 non-null    object
dtypes: object(1)
memory usage: 9.1+ KB
<class 'pandas.core.frame.DataFrame'>
MultiIndex: 400 entries, ('066570.KS', Timestamp('2018-12-31 00:00:00')) to ('YUMC.K', Timestamp('2021-12-31 00:00:00'))
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Capex Data  400 non-null    float64
 1   EBIT Data   400 non-null    float64
 2   EPS Data    400 non-null    float64
 3   OE Data     400 non-null    float64
 4   ROA Data    400 non-null    float64
dtypes: float64(5)
memory usage: 21.6+ KB
<class 'pandas.core.frame.DataFrame'>
MultiIndex: 400 entries, ('066570.KS', Timestamp('2018-12-31 00:00:00')) to ('YUMC.K', Timestamp('2021-

In [ ]:
# sort values
ti_panel=ti_panel.sort_values(['id', 'Dates'], ascending=[True, True])
fin_panel=fin_panel.sort_values(['id', 'Dates'], ascending=[True, True])
emission_panel=emission_panel.sort_values(['id', 'Dates'], ascending=[True, True])

In [ ]:
# combine them into a full panel
panel=pd.concat([ti_panel,fin_panel,emission_panel], axis=1)
panel=panel.sort_values(['id', 'Dates'], ascending=[True, True])
panel.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 404 entries, ('066570.KS', Timestamp('2018-12-31 00:00:00')) to ('YUMC.K', Timestamp('2021-12-31 00:00:00'))
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Sector            400 non-null    object 
 1   Capex Data        400 non-null    float64
 2   EBIT Data         400 non-null    float64
 3   EPS Data          400 non-null    float64
 4   OE Data           400 non-null    float64
 5   ROA Data          400 non-null    float64
 6   Scope 1           400 non-null    float64
 7   Scope 2 Location  400 non-null    float64
 8   Scope 2 Market    400 non-null    float64
dtypes: float64(8), object(1)
memory usage: 34.4+ KB


In [ ]:
panel

Sector    Capex Data     EBIT Data  \
id        Dates                                                            
066570.KS 2018-12-31  Consumer Discretionary  3.219314e+12  2.468549e+12   
          2019-12-31  Consumer Discretionary  3.844371e+12  2.703291e+12   
          2020-12-31  Consumer Discretionary  2.544599e+12  2.436139e+12   
          2021-12-31  Consumer Discretionary  3.083042e+12  3.905108e+12   
1928.T    2018-12-31  Consumer Discretionary  9.458800e+10  1.841640e+11   
...                                      ...           ...           ...   
WOR.AX    2021-12-31                  Energy  6.400000e+07  6.350000e+08   
YUMC.K    2018-12-31  Consumer Discretionary  4.150000e+08  7.610000e+08   
          2019-12-31  Consumer Discretionary  4.700000e+08  8.700000e+08   
          2020-12-31  Consumer Discretionary  4.350000e+08  9.260000e+08   
          2021-12-31  Consumer Discretionary  4.190000e+08  7.020000e+08   

                      EPS Data       OE Data  ROA Data   Scope 1  \
id        Dates                                                    
066570.KS 2018-12-31   422.000  5.402927e+13     0.207  785312.0   
          2019-12-31  9579.000  5.892774e+13     4.365  611675.0   
          2020-12-31  6882.000  5.863837e+13     2.899  440166.0   
          2021-12-31   169.000  5.987004e+13     0.280  332724.0   
1928.T    2018-12-31   175.480  0.000000e+00     9.100   72967.0   
...                        ...           ...       ...       ...   
WOR.AX    2021-12-31     0.825  1.241300e+10     3.892   22238.0   
YUMC.K    2018-12-31     1.280  6.112000e+09    13.630       0.0   
          2019-12-31     1.420  6.359000e+09    14.120  114720.0   
          2020-12-31     1.530  7.462000e+09    13.660  135615.0   
          2021-12-31     1.880  7.850000e+09    12.610  158887.0   

                      Scope 2 Location  Scope 2 Market  
id        Dates                                         
066570.KS 2018-12-31          846510.0          5226.0  
          2019-12-31          891246.0          9114.0  
          2020-12-31          854054.0          5080.0  
          2021-12-31          847031.0        819240.0  
1928.T    2018-12-31           59606.0         58259.0  
...                                ...             ...  
WOR.AX    2021-12-31           32251.0         25973.0  
YUMC.K    2018-12-31               0.0             0.0  
          2019-12-31         1956072.0             0.0  
          2020-12-31         1873757.0             0.0  
          2021-12-31         2158856.0             0.0  

[404 rows x 9 columns]

In [ ]:
panel.to_parquet('full_panel.parquet')

In [ ]:
import os
import pandas as pd
import numpy as np

In [ ]:
# load the full panel
# check the info
panel = pd.read_parquet('full_panel.parquet')
panel.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 404 entries, ('066570.KS', Timestamp('2018-12-31 00:00:00')) to ('YUMC.K', Timestamp('2021-12-31 00:00:00'))
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Sector            400 non-null    object 
 1   Capex Data        400 non-null    float64
 2   EBIT Data         400 non-null    float64
 3   EPS Data          400 non-null    float64
 4   OE Data           400 non-null    float64
 5   ROA Data          400 non-null    float64
 6   Scope 1           400 non-null    float64
 7   Scope 2 Location  400 non-null    float64
 8   Scope 2 Market    400 non-null    float64
dtypes: float64(8), object(1)
memory usage: 34.4+ KB


In [ ]:
# rename columns' name: replace space with underscore
panel.rename(columns={'Scope 1':'Scope_1','Scope 2 Location':'Scope_2_Location','Scope 2 Market':'Scope_2_Market', 'EBIT Data':'EBIT_Data', 'EPS Data':'EPS_Data', 'OE Data':'OE_Data','ROA Data':'ROA_Data','Capex Data':'Capex_Data'},inplace=True)
panel.columns

Index(['Sector', 'Capex_Data', 'EBIT_Data', 'EPS_Data', 'OE_Data', 'ROA_Data',
       'Scope_1', 'Scope_2_Location', 'Scope_2_Market'],
      dtype='object')

In [ ]:
#  add new variables
# sum of emissions: scope 1 +  scope 2 market
# sum of emissions: scope 1 +  scope 2 location
panel['Sum_Emissions_MKT'] = panel['Scope_1'] + panel['Scope_2_Market']
panel['Sum_Emissions_LOC'] = panel['Scope_1'] + panel['Scope_2_Location']
panel.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 404 entries, ('066570.KS', Timestamp('2018-12-31 00:00:00')) to ('YUMC.K', Timestamp('2021-12-31 00:00:00'))
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Sector             400 non-null    object 
 1   Capex_Data         400 non-null    float64
 2   EBIT_Data          400 non-null    float64
 3   EPS_Data           400 non-null    float64
 4   OE_Data            400 non-null    float64
 5   ROA_Data           400 non-null    float64
 6   Scope_1            400 non-null    float64
 7   Scope_2_Location   400 non-null    float64
 8   Scope_2_Market     400 non-null    float64
 9   Sum_Emissions_MKT  400 non-null    float64
 10  Sum_Emissions_LOC  400 non-null    float64
dtypes: float64(10), object(1)
memory usage: 40.7+ KB


In [ ]:
# get descriptive stats
stats=panel.describe()
stats

Capex_Data     EBIT_Data     EPS_Data       OE_Data    ROA_Data  \
count  4.000000e+02  4.000000e+02   400.000000  4.000000e+02  400.000000   
mean   3.957374e+10  4.030810e+10    52.829698  6.353800e+11    5.793165   
std    3.204087e+11  2.941245e+11   589.760742  5.768396e+12    4.697876   
min   -1.060000e+09 -1.638000e+09   -10.600000  0.000000e+00   -5.340000   
25%    2.048000e+08  4.239595e+08     0.823000  3.507862e+08    2.352750   
50%    5.501500e+08  1.216000e+09     1.997500  4.943582e+09    5.050000   
75%    2.127750e+09  4.676500e+09     5.392500  1.821975e+10    8.612500   
max    3.844371e+12  3.905108e+12  9579.000000  5.987004e+13   21.770000   

            Scope_1  Scope_2_Location  Scope_2_Market  Sum_Emissions_MKT  \
count  4.000000e+02      4.000000e+02    4.000000e+02       4.000000e+02   
mean   5.009077e+06      6.934275e+05    4.217503e+05       5.430827e+06   
std    1.642868e+07      1.480465e+06    1.235310e+06       1.665504e+07   
min    0.000000e+00      0.000000e+00    0.000000e+00       0.000000e+00   
25%    1.251950e+04      3.935175e+04    0.000000e+00       4.133525e+04   
50%    1.100000e+05      2.051515e+05    3.539500e+04       2.007685e+05   
75%    6.937168e+05      6.551752e+05    2.600310e+05       1.352187e+06   
max    1.200000e+08      1.305735e+07    1.202208e+07       1.200000e+08   

       Sum_Emissions_LOC  
count       4.000000e+02  
mean        5.702504e+06  
std         1.686734e+07  
min         0.000000e+00  
25%         6.184075e+04  
50%         3.693500e+05  
75%         1.762403e+06  
max         1.250000e+08

In [ ]:
# round to 4 decimal places
stats=stats.round(4)
stats

Capex_Data     EBIT_Data   EPS_Data       OE_Data  ROA_Data  \
count  4.000000e+02  4.000000e+02   400.0000  4.000000e+02  400.0000   
mean   3.957374e+10  4.030810e+10    52.8297  6.353800e+11    5.7932   
std    3.204087e+11  2.941245e+11   589.7607  5.768396e+12    4.6979   
min   -1.060000e+09 -1.638000e+09   -10.6000  0.000000e+00   -5.3400   
25%    2.048000e+08  4.239595e+08     0.8230  3.507862e+08    2.3528   
50%    5.501500e+08  1.216000e+09     1.9975  4.943582e+09    5.0500   
75%    2.127750e+09  4.676500e+09     5.3925  1.821975e+10    8.6125   
max    3.844371e+12  3.905108e+12  9579.0000  5.987004e+13   21.7700   

            Scope_1  Scope_2_Location  Scope_2_Market  Sum_Emissions_MKT  \
count  4.000000e+02      4.000000e+02    4.000000e+02       4.000000e+02   
mean   5.009077e+06      6.934275e+05    4.217503e+05       5.430827e+06   
std    1.642868e+07      1.480465e+06    1.235310e+06       1.665504e+07   
min    0.000000e+00      0.000000e+00    0.000000e+00       0.000000e+00   
25%    1.251950e+04      3.935175e+04    0.000000e+00       4.133525e+04   
50%    1.100000e+05      2.051515e+05    3.539500e+04       2.007685e+05   
75%    6.937168e+05      6.551752e+05    2.600310e+05       1.352187e+06   
max    1.200000e+08      1.305735e+07    1.202208e+07       1.200000e+08   

       Sum_Emissions_LOC  
count       4.000000e+02  
mean        5.702504e+06  
std         1.686734e+07  
min         0.000000e+00  
25%         6.184075e+04  
50%         3.693500e+05  
75%         1.762403e+06  
max         1.250000e+08

In [ ]:
pip install linearmodels

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 18.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 82.3/82.3 kB 6.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.8/43.8 kB 3.8 MB/s eta 0:00:00


In [ ]:
import statsmodels.api as sm
from linearmodels.panel import PanelOLS

In [ ]:
# create Xs
panel.columns

Index(['Sector', 'Capex_Data', 'EBIT_Data', 'EPS_Data', 'OE_Data', 'ROA_Data',
       'Scope_1', 'Scope_2_Location', 'Scope_2_Market', 'Sum_Emissions_MKT',
       'Sum_Emissions_LOC'],
      dtype='object')

In [ ]:
x_vars = ['Capex_Data', 'EBIT_Data', 'EPS_Data', 'OE_Data', 'ROA_Data']

In [ ]:
X = sm.add_constant(panel[x_vars])
X

const    Capex_Data     EBIT_Data  EPS_Data  \
id        Dates                                                     
066570.KS 2018-12-31    1.0  3.219314e+12  2.468549e+12   422.000   
          2019-12-31    1.0  3.844371e+12  2.703291e+12  9579.000   
          2020-12-31    1.0  2.544599e+12  2.436139e+12  6882.000   
          2021-12-31    1.0  3.083042e+12  3.905108e+12   169.000   
1928.T    2018-12-31    1.0  9.458800e+10  1.841640e+11   175.480   
...                     ...           ...           ...       ...   
WOR.AX    2021-12-31    1.0  6.400000e+07  6.350000e+08     0.825   
YUMC.K    2018-12-31    1.0  4.150000e+08  7.610000e+08     1.280   
          2019-12-31    1.0  4.700000e+08  8.700000e+08     1.420   
          2020-12-31    1.0  4.350000e+08  9.260000e+08     1.530   
          2021-12-31    1.0  4.190000e+08  7.020000e+08     1.880   

                           OE_Data  ROA_Data  
id        Dates                               
066570.KS 2018-12-31  5.402927e+13     0.207  
          2019-12-31  5.892774e+13     4.365  
          2020-12-31  5.863837e+13     2.899  
          2021-12-31  5.987004e+13     0.280  
1928.T    2018-12-31  0.000000e+00     9.100  
...                            ...       ...  
WOR.AX    2021-12-31  1.241300e+10     3.892  
YUMC.K    2018-12-31  6.112000e+09    13.630  
          2019-12-31  6.359000e+09    14.120  
          2020-12-31  7.462000e+09    13.660  
          2021-12-31  7.850000e+09    12.610  

[404 rows x 6 columns]

In [ ]:
# create Y df
Y = panel[['Scope_1']]
Y

Scope_1
id        Dates               
066570.KS 2018-12-31  785312.0
          2019-12-31  611675.0
          2020-12-31  440166.0
          2021-12-31  332724.0
1928.T    2018-12-31   72967.0
...                        ...
WOR.AX    2021-12-31   22238.0
YUMC.K    2018-12-31       0.0
          2019-12-31  114720.0
          2020-12-31  135615.0
          2021-12-31  158887.0

[404 rows x 1 columns]

In [ ]:
# specify the model
mod = PanelOLS(Y, X)
res = mod.fit()
print(res)

                          PanelOLS Estimation Summary                           
Dep. Variable:                Scope_1   R-squared:                        0.0817
Estimator:                   PanelOLS   R-squared (Between):              0.1012
No. Observations:                 396   R-squared (Within):              -0.6815
Date:                Fri, Jun 16 2023   R-squared (Overall):              0.0817
Time:                        07:46:01   Log-likelihood                   -7125.7
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      6.9433
Entities:                          99   P-value                           0.0000
Avg Obs:                       4.0000   Distribution:                   F(5,390)
Min Obs:                       4.0000                                           
Max Obs:                       4.0000   F-statistic (robust):             6.9433
                            

/usr/local/lib/python3.10/dist-packages/linearmodels/panel/model.py:1214: MissingValueWarning: 
Inputs contain missing values. Dropping rows with missing observations.
  super().__init__(dependent, exog, weights=weights, check_rank=check_rank)


It's worth noting that the p-values for const, Capex Data, EBIT Data, EPS Data, and OE Data are relatively high, indicating that these variables may not be statistically significant in explaining the variation in the dependent variable. However, the p-value for ROA Data is very low (p < 0.001), suggesting that it is a statistically significant variable.

The F-statistic is 6.9433, with a corresponding p-value of 0.0000, indicating that at least one of the independent variables in the model is jointly significant.

Overall, this summary provides an overview of the regression results and the statistical significance of the variables in explaining the variation in the dependent variable.

This means that, holding other variables constant, an increase in these independent variables is associated with a decrease in the dependent variable (Scope_1).

Specifically:

An increase in Capex Data by one unit is associated with a decrease in Scope_1 by -5.356e-06 units.
An increase in EBIT Data by one unit is associated with a decrease in Scope_1 by -1.093e-05 units.
An increase in EPS Data by one unit is associated with a decrease in Scope_1 by -436.55 units.
An increase in ROA Data by one unit is associated with a decrease in Scope_1 by -1.014e+06 units.
However, the parameter estimate for OE Data is positive, indicating that an increase in OE Data by one unit is associated with an increase in Scope_1 by 7.203e-07 units.



PanelOLS

In [ ]:
# REGRESSIONS MODEL 1
mod1 = PanelOLS.from_formula('Scope_1 ~ BS_TOT_ASSET + CASH_TO_TOT_ASSET + EBIT + ESG_Gov + HISTORICAL_MARKET_CAP', data = panel)
results1 = mod1.fit(cov_type='clustered', cluster_entity=True)
print(results1)

In [ ]:
# REGRESSIONS MODEL 1
mod1 = PanelOLS.from_formula('Scope_1 ~ Capex_Data + EBIT_Data + EPS_Data + OE_Data + ROA_Data', data = panel)
results1 = mod1.fit(cov_type='clustered', cluster_entity=True)
print(results1)

                          PanelOLS Estimation Summary                           
Dep. Variable:                Scope_1   R-squared:                        0.0038
Estimator:                   PanelOLS   R-squared (Between):              0.0041
No. Observations:                 396   R-squared (Within):              -0.0061
Date:                Fri, Jun 16 2023   R-squared (Overall):              0.0038
Time:                        06:10:30   Log-likelihood                   -7159.6
Cov. Estimator:             Clustered                                           
                                        F-statistic:                      0.3003
Entities:                          99   P-value                           0.9125
Avg Obs:                       4.0000   Distribution:                   F(5,391)
Min Obs:                       4.0000                                           
Max Obs:                       4.0000   F-statistic (robust):             7.1706
                            

/usr/local/lib/python3.10/dist-packages/linearmodels/panel/model.py:1214: MissingValueWarning: 
Inputs contain missing values. Dropping rows with missing observations.
  super().__init__(dependent, exog, weights=weights, check_rank=check_rank)


Capex_Data: The coefficient estimate is 2.318e-06. Holding other variables constant, an increase in Capex_Data by one unit is associated with an increase in Scope_1 by 2.318e-06 units. However, the p-value (0.2954) suggests that this relationship is not statistically significant.

EBIT_Data: The coefficient estimate is 1.83e-07. Holding other variables constant, an increase in EBIT_Data by one unit is associated with an increase in Scope_1 by 1.83e-07 units. The p-value (0.9311) indicates that this relationship is not statistically significant.

EPS_Data: The coefficient estimate is -115.24. Holding other variables constant, an increase in EPS_Data by one unit is associated with a decrease in Scope_1 by 115.24 units. However, the p-value (0.5820) suggests that this relationship is not statistically significant.

OE_Data: The coefficient estimate is -1.203e-07. Holding other variables constant, an increase in OE_Data by one unit is associated with a decrease in Scope_1 by 1.203e-07 units. The p-value (0.4768) indicates that this relationship is not statistically significant.

ROA_Data: The coefficient estimate is 1.435e+05. Holding other variables constant, an increase in ROA_Data by one unit is associated with an increase in Scope_1 by 1.435e+05 units. The p-value (0.0016) suggests that this relationship is statistically significant.

In summary, based on the provided results, none of the variables (except ROA_Data) show statistically significant relationships with the dependent variable (Scope_1). The R-squared values indicate that the independent variables explain only a small proportion of the variance in Scope_1.

In [ ]:
# REGRESSIONS MODEL 2
mod2 = PanelOLS.from_formula('Scope_2_Market ~ Capex_Data + EBIT_Data + EPS_Data + OE_Data + ROA_Data', data = panel)
results2 = mod2.fit(cov_type='clustered', cluster_entity=True)
print(results2)

                          PanelOLS Estimation Summary                           
Dep. Variable:         Scope_2_Market   R-squared:                        0.0421
Estimator:                   PanelOLS   R-squared (Between):              0.0488
No. Observations:                 396   R-squared (Within):              -0.0438
Date:                Fri, Jun 16 2023   R-squared (Overall):              0.0421
Time:                        06:17:16   Log-likelihood                   -6131.4
Cov. Estimator:             Clustered                                           
                                        F-statistic:                      3.4338
Entities:                          99   P-value                           0.0048
Avg Obs:                       4.0000   Distribution:                   F(5,391)
Min Obs:                       4.0000                                           
Max Obs:                       4.0000   F-statistic (robust):             13.068
                            

/usr/local/lib/python3.10/dist-packages/linearmodels/panel/model.py:1214: MissingValueWarning: 
Inputs contain missing values. Dropping rows with missing observations.
  super().__init__(dependent, exog, weights=weights, check_rank=check_rank)


Capex_Data: The coefficient estimate is -3.103e-07. Holding other variables constant, an increase in Capex_Data by one unit is associated with a decrease in Scope_2_Market by 3.103e-07 units. The p-value (0.4289) suggests that this relationship is not statistically significant.

EBIT_Data: The coefficient estimate is 6.467e-07. Holding other variables constant, an increase in EBIT_Data by one unit is associated with an increase in Scope_2_Market by 6.467e-07 units. The p-value (0.0261) indicates that this relationship is statistically significant.

EPS_Data: The coefficient estimate is -22.581. Holding other variables constant, an increase in EPS_Data by one unit is associated with a decrease in Scope_2_Market by 22.581 units. The p-value (0.5320) suggests that this relationship is not statistically significant.

OE_Data: The coefficient estimate is -9.73e-09. Holding other variables constant, an increase in OE_Data by one unit is associated with a decrease in Scope_2_Market by 9.73e-09 units. The p-value (0.7669) indicates that this relationship is not statistically significant.

ROA_Data: The coefficient estimate is 3.557e+04. Holding other variables constant, an increase in ROA_Data by one unit is associated with an increase in Scope_2_Market by 3.557e+04 units. The p-value (0.0125) suggests that this relationship is statistically significant.

Overall, the regression model explains a small proportion of the variance in Scope_2_Market, as indicated by the low R-squared values. Among the independent variables, only EBIT_Data and ROA_Data show statistically significant relationships with Scope_2_Market based on their respective p-values. The other variables (Capex_Data, EPS_Data, and OE_Data) do not appear to be statistically significant in explaining the variation in Scope_2_Market.

In [ ]:
#  REGRESSIONS MODEL 3
mod3 = PanelOLS.from_formula('Scope_2_Location ~ Capex_Data + EBIT_Data + EPS_Data + OE_Data + ROA_Data', data = panel)
results3 = mod3.fit(cov_type='clustered', cluster_entity=True)
print(results3)

                          PanelOLS Estimation Summary                           
Dep. Variable:       Scope_2_Location   R-squared:                        0.0704
Estimator:                   PanelOLS   R-squared (Between):              0.0787
No. Observations:                 396   R-squared (Within):              -0.0953
Date:                Fri, Jun 16 2023   R-squared (Overall):              0.0704
Time:                        06:20:49   Log-likelihood                   -6214.1
Cov. Estimator:             Clustered                                           
                                        F-statistic:                      5.9253
Entities:                          99   P-value                           0.0000
Avg Obs:                       4.0000   Distribution:                   F(5,391)
Min Obs:                       4.0000                                           
Max Obs:                       4.0000   F-statistic (robust):             1463.8
                            

/usr/local/lib/python3.10/dist-packages/linearmodels/panel/model.py:1214: MissingValueWarning: 
Inputs contain missing values. Dropping rows with missing observations.
  super().__init__(dependent, exog, weights=weights, check_rank=check_rank)


Capex_Data: The coefficient estimate is 9.246e-09. Holding other variables constant, an increase in Capex_Data by one unit is associated with an increase in Scope_2_Location by 9.246e-09 units. The p-value (0.9826) suggests that this relationship is not statistically significant.

EBIT_Data: The coefficient estimate is 5.824e-08. Holding other variables constant, an increase in EBIT_Data by one unit is associated with an increase in Scope_2_Location by 5.824e-08 units. The p-value (0.8705) indicates that this relationship is not statistically significant.

EPS_Data: The coefficient estimate is -43.031. Holding other variables constant, an increase in EPS_Data by one unit is associated with a decrease in Scope_2_Location by 43.031 units. The p-value (0.3314) suggests that this relationship is not statistically significant.

OE_Data: The coefficient estimate is 1.437e-08. Holding other variables constant, an increase in OE_Data by one unit is associated with an increase in Scope_2_Location by 1.437e-08 units. The p-value (0.6870) indicates that this relationship is not statistically significant.

ROA_Data: The coefficient estimate is 5.786e+04. Holding other variables constant, an increase in ROA_Data by one unit is associated with an increase in Scope_2_Location by 5.786e+04 units. The p-value (0.0015) suggests that this relationship is statistically significant.

Overall, the regression model explains a moderate proportion of the variance in Scope_2_Location, as indicated by the R-squared value. Among the independent variables, only ROA_Data shows a statistically significant relationship with Scope_2_Location based on its p-value. The other variables (Capex_Data, EBIT_Data, EPS_Data, and OE_Data) do not appear to be statistically significant in explaining the variation in Scope_2_Location.

In [ ]:
#  REGRESSIONS MODEL 4
mod4 = PanelOLS.from_formula('Sum_Emissions_MKT ~ Capex_Data + EBIT_Data + EPS_Data + OE_Data + ROA_Data', data = panel)
results4 = mod4.fit(cov_type='clustered', cluster_entity=True)
print(results4)

                          PanelOLS Estimation Summary                           
Dep. Variable:      Sum_Emissions_MKT   R-squared:                        0.0057
Estimator:                   PanelOLS   R-squared (Between):              0.0061
No. Observations:                 396   R-squared (Within):              -0.0116
Date:                Fri, Jun 16 2023   R-squared (Overall):              0.0057
Time:                        06:22:25   Log-likelihood                   -7167.1
Cov. Estimator:             Clustered                                           
                                        F-statistic:                      0.4499
Entities:                          99   P-value                           0.8133
Avg Obs:                       4.0000   Distribution:                   F(5,391)
Min Obs:                       4.0000                                           
Max Obs:                       4.0000   F-statistic (robust):             6.0572
                            

/usr/local/lib/python3.10/dist-packages/linearmodels/panel/model.py:1214: MissingValueWarning: 
Inputs contain missing values. Dropping rows with missing observations.
  super().__init__(dependent, exog, weights=weights, check_rank=check_rank)


Capex_Data: The coefficient estimate is 2.008e-06. Holding other variables constant, an increase in Capex_Data by one unit is associated with an increase in Sum_Emissions_MKT by 2.008e-06 units. The p-value (0.3534) suggests that this relationship is not statistically significant.

EBIT_Data: The coefficient estimate is 8.298e-07. Holding other variables constant, an increase in EBIT_Data by one unit is associated with an increase in Sum_Emissions_MKT by 8.298e-07 units. The p-value (0.7040) indicates that this relationship is not statistically significant.

EPS_Data: The coefficient estimate is -137.82. Holding other variables constant, an increase in EPS_Data by one unit is associated with a decrease in Sum_Emissions_MKT by 137.82 units. The p-value (0.5353) suggests that this relationship is not statistically significant.

OE_Data: The coefficient estimate is -1.3e-07. Holding other variables constant, an increase in OE_Data by one unit is associated with a decrease in Sum_Emissions_MKT by 1.3e-07 units. The p-value (0.4406) indicates that this relationship is not statistically significant.

ROA_Data: The coefficient estimate is 1.791e+05. Holding other variables constant, an increase in ROA_Data by one unit is associated with an increase in Sum_Emissions_MKT by 1.791e+05 units. The p-value (0.0005) suggests that this relationship is statistically significant.

Overall, the regression model explains a very small proportion of the variance in Sum_Emissions_MKT, as indicated by the low R-squared value. Among the independent variables, only ROA_Data shows a statistically significant relationship with Sum_Emissions_MKT based on its p-value. The other variables (Capex_Data, EBIT_Data, EPS_Data, and OE_Data) do not appear to be statistically significant in explaining the variation in Sum_Emissions_MKT.

In [ ]:
#  REGRESSIONS MODEL 5
mod5 = PanelOLS.from_formula('Sum_Emissions_LOC ~ Capex_Data + EBIT_Data + EPS_Data + OE_Data + ROA_Data', data = panel)
results5 = mod5.fit(cov_type='clustered', cluster_entity=True)
print(results5)

                          PanelOLS Estimation Summary                           
Dep. Variable:      Sum_Emissions_LOC   R-squared:                        0.0070
Estimator:                   PanelOLS   R-squared (Between):              0.0075
No. Observations:                 396   R-squared (Within):              -0.0135
Date:                Fri, Jun 16 2023   R-squared (Overall):              0.0070
Time:                        06:23:14   Log-likelihood                   -7173.3
Cov. Estimator:             Clustered                                           
                                        F-statistic:                      0.5512
Entities:                          99   P-value                           0.7374
Avg Obs:                       4.0000   Distribution:                   F(5,391)
Min Obs:                       4.0000                                           
Max Obs:                       4.0000   F-statistic (robust):             16.574
                            

/usr/local/lib/python3.10/dist-packages/linearmodels/panel/model.py:1214: MissingValueWarning: 
Inputs contain missing values. Dropping rows with missing observations.
  super().__init__(dependent, exog, weights=weights, check_rank=check_rank)


Capex_Data: The coefficient estimate is 2.327e-06. Holding other variables constant, an increase in Capex_Data by one unit is associated with an increase in Sum_Emissions_LOC by 2.327e-06 units. The p-value (0.3014) suggests that this relationship is not statistically significant.

EBIT_Data: The coefficient estimate is 2.413e-07. Holding other variables constant, an increase in EBIT_Data by one unit is associated with an increase in Sum_Emissions_LOC by 2.413e-07 units. The p-value (0.9154) indicates that this relationship is not statistically significant.

EPS_Data: The coefficient estimate is -158.27. Holding other variables constant, an increase in EPS_Data by one unit is associated with a decrease in Sum_Emissions_LOC by 158.27 units. The p-value (0.4947) suggests that this relationship is not statistically significant.

OE_Data: The coefficient estimate is -1.059e-07. Holding other variables constant, an increase in OE_Data by one unit is associated with a decrease in Sum_Emissions_LOC by 1.059e-07 units. The p-value (0.5478) indicates that this relationship is not statistically significant.

ROA_Data: The coefficient estimate is 2.014e+05. Holding other variables constant, an increase in ROA_Data by one unit is associated with an increase in Sum_Emissions_LOC by 2.014e+05 units. The p-value (0.0002) suggests that this relationship is statistically significant.

Overall, the regression model explains a very small proportion of the variance in Sum_Emissions_LOC, as indicated by the low R-squared value. Among the independent variables, only ROA_Data shows a statistically significant relationship with Sum_Emissions_LOC based on its p-value. The other variables (Capex_Data, EBIT_Data, EPS_Data, and OE_Data) do not appear to be statistically significant in explaining the variation in Sum_Emissions_LOC.

In [ ]:
from linearmodels.panel.results import compare

In [ ]:
print(compare({"Model 1": results1, "Model 2": results2, "Model 3": results3}, stars=True))

                            Model Comparison                            
                             Model 1          Model 2            Model 3
------------------------------------------------------------------------
Dep. Variable                Scope_1   Scope_2_Market   Scope_2_Location
Estimator                   PanelOLS         PanelOLS           PanelOLS
No. Observations                 396              396                396
Cov. Est.                  Clustered        Clustered          Clustered
R-squared                     0.0038           0.0421             0.0704
R-Squared (Within)           -0.0061          -0.0438            -0.0953
R-Squared (Between)           0.0041           0.0488             0.0787
R-Squared (Overall)           0.0038           0.0421             0.0704
F-statistic                   0.3003           3.4338             5.9253
P-value (F-stat)              0.9125           0.0048             0.0000
===================== ============== ==============

Model 3 has the highest R-squared value among the three models, indicating that it explains a greater proportion of the variance in the dependent variable compared to the other models.
Model 2 has a moderately higher R-squared value than Model 1, suggesting that it provides a better fit for explaining Scope_2_Market emissions.
Model 3 has a higher R-squared value than both Model 1 and Model 2, indicating a better fit for explaining Scope_2_Location emissions.
Among the independent variables, ROA consistently appears as a significant predictor of emissions in both Model 2 and Model 3, suggesting its potential importance in understanding emissions scopes.
The other independent variables do not consistently show significant associations with the emissions scopes across the models.
In summary, Model 3 appears to be the most effective in explaining emissions related to Scope_2_Location, while Model 2 provides a modest level of explanation for emissions related to Scope_2_Market. Model 1, on the other hand, does not demonstrate a strong relationship between the independent variables and Scope_1 emissions.

In [ ]:
print(compare({"Model 1": results1, "Model 2": results2, "Model 3": results3, "Model 4": results4, "Model 5": results5}, stars=True))

                                                Model Comparison                                                
                             Model 1          Model 2            Model 3             Model 4             Model 5
----------------------------------------------------------------------------------------------------------------
Dep. Variable                Scope_1   Scope_2_Market   Scope_2_Location   Sum_Emissions_MKT   Sum_Emissions_LOC
Estimator                   PanelOLS         PanelOLS           PanelOLS            PanelOLS            PanelOLS
No. Observations                 396              396                396                 396                 396
Cov. Est.                  Clustered        Clustered          Clustered           Clustered           Clustered
R-squared                     0.0038           0.0421             0.0704              0.0057              0.0070
R-Squared (Within)           -0.0061          -0.0438            -0.0953             -0.0116    

1. Model 1 does not provide strong evidence of a significant relationship between the independent variables and Scope_1 emissions.
2. Model 2 suggests that EBIT and ROA may be important predictors of Scope_2_Market emissions.
3. Model 3 suggests that ROA may be a key factor in understanding Scope_2_Location emissions.
4. Model 4 does not provide strong evidence of a significant relationship between the independent variables and Sum_Emissions_MKT.
5. Model 5 suggests that ROA may be a relevant factor in understanding Sum_Emissions_LOC.

In summary, Model 3 provides the highest level of explanation for Scope_2_Location emissions, while Model 2 shows a modest level of explanation for Scope_2_Market emissions. Models 1 and 4 do not demonstrate strong relationships between the independent variables and their respective dependent variables. Model 5 suggests that ROA may be relevant in understanding Sum_Emissions_LOC. Overall, ROA consistently emerges as a potentially important factor in explaining emissions across some of the models.